solver/src/analytics.rs

```diff
@@ -0,0 +1,159 @@
+use crate::settlement::Settlement;
+use bigdecimal::Zero;
+use model::order::OrderUid;
+use num::BigRational;
+use shared::conversions::U256Ext;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+
+/// Record metric with surplus achieved in winning settlement
+/// vs that which was unrealized in other feasible solutions.
+pub fn num_unsettled_orders_with_better_surplus(
+    submitted: &Settlement,
+    all: impl Iterator<Item = Settlement>,
+) -> usize {
+    let submitted_prices = submitted
+        .clearing_prices()
+        .iter()
+        .map(|(token, price)| (token, price.to_big_rational()))
+        .collect::<HashMap<_, _>>();
+    let submitted_surplus: HashMap<_, _> = submitted
+        .trades()
+        .iter()
+        .map(|trade| {
+            (
+                trade.order.order_meta_data.uid,
+                trade
+                    .surplus(
+                        &submitted_prices[&trade.order.order_creation.sell_token],
+                        &submitted_prices[&trade.order.order_creation.buy_token],
+                    )
+                    .unwrap_or_else(BigRational::zero),
+            )
+        })
+        .collect();
+
+    let mut best_surplus: HashMap<OrderUid, BigRational> = HashMap::new();
+    for settlement in all {
+        let trades = settlement.trades();
+        let clearing_prices = settlement
+            .clearing_prices()
+            .iter()
+            .map(|(token, price)| (token, price.to_big_rational()))
+            .collect::<HashMap<_, _>>();
+
+        for trade in trades {
+            let order_id = trade.order.order_meta_data.uid;
+            if submitted_surplus.contains_key(&order_id) {
+                let surplus = trade
+                    .surplus(
+                        &clearing_prices[&trade.order.order_creation.sell_token],
+                        &clearing_prices[&trade.order.order_creation.buy_token],
+                    )
+                    .unwrap_or_else(BigRational::zero);
+                let entry = best_surplus.entry(order_id);
+                match entry {
+                    Entry::Occupied(mut entry) => {
+                        let value = entry.get_mut();
+                        *value = (value.clone()).max(surplus);
+                    }
+                    Entry::Vacant(entry) => {
+                        entry.insert(surplus);
+                    }
+                }
+            }
+        }
+    }
+
+    let mut count = 0;
+    for (order_id, submitted) in submitted_surplus.iter() {
+        let best = best_surplus.get(order_id).expect("exists by construction");
+        if best > &(BigRational::new(101.into(), 100.into()) * submitted) {
```

Comment:
Yes. You mean the 1% right. I completely agree. I can add this simple thing in a follow up unless anyone insists it happen here. Other wise I'll leave a todo with an issue.

Reply:
https://github.com/gnosis/gp-v2-services/issues/1350

Reply:
Maybe, since we plan on relying on metrics more than logs, we can just log all cases and not miss cases that we wanted to know about (for example 0.9% on a 10M USD trade is 9K USD and wouldn't get logged).

Reply:
I think you may have missed the comment :point_up:

Reply:
I will log whenever the best alternative is better and in #1350 we can configure when to error instead!